include/uart.h

```diff
@@ -28,6 +29,144 @@ extern "C" {
 
 #include <device.h>
 
+/**
+ * @brief UART controller configuration structure
+ *
+ * @note Used for UART high level API
+ *
+ * @param baudrate   Baudrate setting.
+ * @param rx_timeout RX timeout in ms.
+ * @param flags      Bit field with following parts:
+ *
+ *     parity       [ 0 : 2 ]    - Parity bits.
+ *     stop_bits    [ 3 : 4 ]    - Stop bits.
+ *     data_bits    [ 5 : 7 ]    - Data bits.
+ *     flow_ctrl    [ 8 : 9 ]    - Flow control setting.
+ */
+struct uart_config {
+	u32_t baudrate;
+	u32_t rx_timeout;
+	u32_t flags;
```

Comment:
How about replace `u32_t flags` with:

```
u8_t parity;
u8_t stop_bits;
u8_t data_bits;
u8_t flow_ctrl;
```

And than have enum values instead of defines.

Reply:
I used enums before, after [conversation with](https://github.com/zephyrproject-rtos/zephyr/pull/10820#discussion_r228126317) @tbursztyka I changed it to bitmasks.
I prefer enums, which one should I finally use?

Reply:
@tbursztyka regarding timeout. I don't think that kconfig is a good place since it will vary between instances.

What do you think about using character length as timeout unit. Main advantage is that if you change baudrate you don't need to change timeout because it will be aligned.

Reply:
Is the savings really worth it?  We're probably adding code size to do shift and masks vs just being able to directly access the variable.  Also, I feels like the shift/mask is error prone to the driver developer from a security point of view.  (I feel like I've seen over and over again coverity reporting simliar issues in GPIO drivers related to some bit masks there).

Reply:
> What do you think about using character length as timeout unit.

Please use standard units (milliseconds) for timeout.

Reply:
> Please use standard units (milliseconds) for timeout.

Why? Because POSIX is doing that? I also got used to using standard units but then eventually i was setting timeout in characters so it is aligned when i change baudrate.

If we use standard units then it should be microseconds since on 1mbaud 1 character is ~10us and I would like to see macro for calculating character length in uart.h (e.g.`UART_CHAR_USEC(baudrate)`)

Reply:
@galak The space being saved could be relevant systems with numerous ports (such case exists in industry). Also, most of the built time pre-computed masks are directly usable and as such will take exact same amount of code as enum comparison. Anyway, to me, it's mostly a question of consistency against other APIs we have.

@nordic-krch What would be the use-case for such unit and macros? Sending 1 character takes around 10µs, but you need to take into account all the cycles necessary to program the controller, sets the buffers, get the int to actually tx or rx etc... Then 1ms as the smallest timeout sounds just right. Could you detail, because to me as well timeouts should be fine in ms. Also, we miss the timeout error here, -EGAIN.

Reply:
@tbursztyka If we have communication like HCI with ACK's and you might have short packets (<10 bytes) which last <100us @1mbaud. There will be turnaround time of 1ms because application will not receive packet until timeout (1ms later) so such long timeout will reduce uart channel capacity. Sending ten 10byte packets will take at least 10ms were actual transfers will take 10% of that. 

We have seen such issues with communication over uart with PC. Timeout on PC side was limiting throughput so much that there was no significant difference between 115200 and 1M.

Reply:
> There will be turnaround time of 1ms because application will not receive packet until timeout

in case of error only though yes.
And, isn't this use case valid only on blocking separate tx/rx calls?

This proposal is also trying to solve full duplex, so you could still tx, while rx would be pending, and you would get tx/rx termination events via cb. Wouldn't that fix your use case without the need for µs timeout?

Just asking, I am not against µs. If that one proves to be more relevant than ms, then fine. However, I am also concerned about driver side: exposed kernel primitives only takes ms (semaphore or else). Maybe the sys_clock.h _timeout then.